### PR TITLE
Introduce NioTransport into framework for testing

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit introduces a nio based tcp transport into framework for
testing.

Currently Elasticsearch uses a simple blocking tcp transport for
testing purposes (MockTcpTransport). This diverges from production
where our current transport (netty) is non-blocking.

The point of this commit is to introduce a testing variant that more
closely matches the behavior of production instances.

### CODE REVIEW

**Comment by tbrooks8** [comment]
I have not modified any of the sites where we currently use MockTcpTransport. I'm not sure if we would want to introduce randomization at different spots for picking which is used (NioTransport or MockTcpTransport)? I've seen that we do that in some different spots.

**Comment by s1monw** [comment]
just 4k SLoC ;) I am trying....

File: test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java]: this is a todo? nocommit? I wonder if we can remove it since it's not an `@Override`?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java]: maybe we can remove this class and use a simple interface for `handleMessage` and `handleException`

**Review by s1monw — CHANGES_REQUESTED** [review]
I made it until `WriteOperation.java` I will continue later! thanks for doing this @tbrooks8

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NioClient.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NioClient.java]: I changed this to be one try / catch. And it catches all RuntimeExceptions and all checked exceptions that are thrown.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/TcpReadHandler.java]: More of like a "possibly todo". Currently it seems like we only do anything with pings on the "send" side. I had added a handler in case we ever wants to trace or keep metrics about or anything from the receive side. I'll think about it a bit but I'll probably remove it.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java]: Do you mean ensure that the channel is open or that the selector is open?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java]: The main reason for the CountDownLatch was to ensure that caller knew that all closing steps had been completed. I modified this to keep the latch internal and to just wait inside of close(). 

That honestly seemed like the simplest and in retrospect there was no reason to leak the latch. Let me know if that seems reasonable to you.

**Comment by tbrooks8** [comment]
Thanks. I've made a lot of modifications based on your review.

I have not yet gotten to javadocs. But I do plan on definitely adding javadocs to: selectors, handlers, and channels.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java]: it's not trivial :) I took a look at it and here is a suggestion, lemme know what you think:

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java b/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
index 22d9193d8e..a88dd280ae 100644
--- a/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
+++ b/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
@@ -19,7 +19,6 @@
 
 package org.elasticsearch.transport.nio;
 
-import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.util.BigArrays;
 import org.elasticsearch.transport.nio.channel.NioChannel;
 
@@ -31,23 +30,21 @@ import java.util.Collections;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
 
 public abstract class ESSelector implements Closeable {
 
-    protected static final int NOT_STARTED = 0;
-    protected static final int RUNNING = 1;
-    protected static final int STOPPED = 2;
 
     protected Selector selector;
-    protected volatile int state = 0;
+    protected final AtomicBoolean isClosed = new AtomicBoolean(false);
     protected final ConcurrentLinkedQueue<NioChannel> channelsToClose = new ConcurrentLinkedQueue<>();
     protected final Set<NioChannel> registeredChannels = Collections.newSetFromMap(new ConcurrentHashMap<NioChannel, Boolean>());
 
     private final EventHandler eventHandler;
     private final BigArrays bigArrays;
     private volatile Thread thread;
-    private CountDownLatch shutdownLatch = new CountDownLatch(1);
+    private final ReentrantLock runLock = new ReentrantLock();
     protected ESSelector(EventHandler eventHandler, BigArrays bigArrays) throws IOException {
         this(eventHandler, bigArrays, Selector.open());
     }
@@ -59,18 +56,28 @@ public abstract class ESSelector implements Closeable {
     }
 
     public void runLoop() {
-        setThread();
-        state = RUNNING;
-        try {
-            while (state == RUNNING) {
-                singleLoop();
-            }
-        } finally {
+        if (runLock.tryLock()) {
             try {
-                cleanup();
+                ensureOpen();
+                setThread();
+                while (isClosed.get() == false) {
+                    singleLoop();
+                }
             } finally {
-                shutdownLatch.countDown();
+                try {
+                    cleanup();
+                } finally {
+                    runLock.unlock();
+                }
             }
+        } else {
+            throw new IllegalStateException("selector is already running");
+        }
+    }
+
+    private void ensureOpen() {
+        if (isClosed.get()) {
+            throw new IllegalStateException("selector is already closed");
         }
     }
 
@@ -79,7 +86,7 @@ public abstract class ESSelector implements Closeable {
             closePendingChannels();
             doSelect(300);
         } catch (ClosedSelectorException e) {
-            if (state == RUNNING) {
+            if (isRunning()) {
                 throw e;
             }
         } catch (IOException e) {
@@ -114,19 +121,12 @@ public abstract class ESSelector implements Closeable {
     }
 
     public void close(boolean shouldInterrupt) throws IOException {
-        int currentState = this.state;
-        state = STOPPED;
-        selector.close();
-        if (shouldInterrupt && thread != null) {
-            thread.interrupt();
-        }
-        if (currentState == NOT_STARTED) {
-            shutdownLatch.countDown();
-        }
-        try {
-            shutdownLatch.await();
-        } catch (InterruptedException e) {
-            throw new ElasticsearchException(e);
+        if (isClosed.compareAndSet(false, true)) {
+            selector.close();
+            if (shouldInterrupt && thread != null) {
+                thread.interrupt();
+            }
+            runLock.lock(); // wait for the shutdown to complete
         }
     }
 
@@ -163,6 +163,6 @@ public abstract class ESSelector implements Closeable {
     }
 
     public boolean isRunning() {
-        return state == RUNNING;
+        return runLock.isLocked();
     }
 }
```

File: test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java]: yeah I added that method in the diff above

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: are these status codes? maybe put their purpuse in the constant name?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: extra newlines?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: I wonder if we can extract and share this from TCP somehow.. can happen before or after this lands...

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: leftover?

Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: should we implement a `ByteBufferBytesReference` ?

**Review by s1monw — CHANGES_REQUESTED** [review]
I did another pass. I will now look into how things play together and debug my way through it. I like it so far..but I could use some docs here and there to understand better :)

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/AbstractNioChannel.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/AbstractNioChannel.java]: At times java generics confuse me, but I do not think you can do this:

`TcpTransport<AbstractNioChannel<S extends SelectableChannel & NetworkChannel>>`

Which means that instead you have to do this:

`TcpTransport<AbstractNioChannel>`

And then you get a bunch of warnings when compiling and stuff. That is why I introduced the interface.

Let me know if you have a solution that can drop the interface and work well with TCPTransport that I am not thinking of.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/AbstractNioChannel.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/AbstractNioChannel.java]: The thread safety in my current access patterns was provided by checked the `state` variable first.

However, I agree that is not immediately apparent + we expose a getter for selector, so I went ahead and made it volatile.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java]: I'm not sure. `finishConnect()` a method that you have to call on the real SocketChannel to ensure that it is ready to read and write. In this case, this method is making that call + doing other things necessary to ensure that our channel abstraction is set up (and futures are completed). 

And we do not really "wait" at any point. If the raw "finishConnect()" method returns false well wait until a OP_CONNECT event to try again.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/CloseFuture.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/CloseFuture.java]: Wait - I thought that you wanted me to clear the interrupt status?

The source of `Thread.interrupted()` is:

```
currentThread().isInterrupted(true);
```

Which clears the interrupt status?

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java]: ok javadocs will help :)

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: This is where I was playing around with using our byte array recycler for the reads. I decided anything like that was not necessary for this PR. I'll remove the comments.

**Comment by tbrooks8** [comment]
@s1monw I'm still working on updates based on your last two reviews. So no need to continue to do another review or anything. But I was wondering if you had the time tomorrow if you might take a look at close() refactorings I made based on your feedback.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java]: I'll change it to one setting. The contexts normally need to take a reference to the channel. So the channel must first be constructed to created the contexts.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpWriteContext.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpWriteContext.java]: The ByteBuffers have internal markers that the low-level flush operations use. So that is why stuff is not written twice. The WriteOperation#decrementRemaining work is so we know when the total WriteOperation (multiple buffers) is complete and can call the listener. 

I don't disagree with your thought about moving writing into WriteOperation though. I'm taking a look at what that might look like.

**Comment by s1monw** [comment]
@tbrooks8 is this ready for another round?

**Comment by tbrooks8** [comment]
Hi @s1monw. Thanks for all the feedback. I've been working on a few other issues recently, so I have not taken a look at this in a while. But here is my assessment of where things are.

1. I have added documentation to the channels, selectors, and event handlers. These are the primary pieces that I thought needed documenation.

2. I looked into just using raw `ActionFuture` / `ListenableActionFuture` opposed to the specialized extensions I created. I found a few issues.

- The `ListenableActionFuture` takes an `ActionListener`. In the case of `onException`, this only allows the passing of the exception to the listener, not a value (in this case the channel to be closed). This means I would need to allocate a closure with a reference to the channel for every channel opposed to just passing the method reference like I do now.

```
nioSocketChannel.getCloseFuture().setListener(openChannels::channelClosed);
```

- In the specialized version I only expose setting `IOException` and `RuntimeException` as a failure. This is opposed to the raw `ActionFuture` which can take any exception. The implication here is that when I check for failures such as in `NioClient` it makes checking the failure reason harder. It wraps things like `IOException` in `UncategorizedException`. Just throwing `UncategorizedException` from `NioClient` opposed to `IOException` causes a test in `SimpleNioTransportTests` to fail (where we are expecting `IOException`). So either the test would need to be changed or I would need to work or unwrapping exceptions as some intermediate step.

- There are nice helper methods on the specialized verions such as: `isClosed`, `awaitClose`, `getCloseException`, `awaitConnectionComplete`, `isConnectComplete`, `connectFailed`. The answers to these methods can be determined by the raw `actionGet` method, but that requires catching a lot different exceptions and replicating logic in different places to determine what happened.

I guess the answer is: we could use the generic futures. And maybe there a few things we could change about the generic futures to make them a little nicer for these tasks. But when I was going through it, there were a lot of changes. And some things that were not super nice that resulted. I'm not sure if this is something we can hold off on and maybe consider as a follow-up?

3. I have not yet made any of your suggested changes to making channel writes in the `WriteOperation` class. Although that is something that I agree is worth exploring.

4. As to your question about `ReadContext` and `WriteContext` interfaces, there are interfaces, because I assume at some point we will want Read and WriteContexts that handle SSL (which will require different logic).

5. I recently reworked the `TcpReadContext` a little. I moved it closer to using our concept on a `BytesReference` opposed to the raw `ByteBuffer`. This was kind of in anticipation for where we might not always know exactly how many bytes we need to read (such as encrypted data). So in this case we would read a buffer, pass it to a ssl decoder, and then to a tcp frame decoder once a ssl packet has been read. The BytesReference abstraction works okay for this. But it is limited without reader and writer indexes. We might need to create a ByteBufferReference (as you suggested) with indexes that keep state for passing around to different decoders. But this can definitely be follow-up work. Or it might come up when I am looking at the `WriteContext` work.` The current solution should work for reading plaintext traffic.

Other than the futures changes and the write context work. I think I have gotten to your requested changes. I think I should be able to take a look at your write context suggestions this week.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/WriteOperation.java
Comment (tbrooks8) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/WriteOperation.java]: The usage of this class spans both nio and nio.channel packages.

**Comment by tbrooks8** [comment]
@s1monw 

I have modified the `WriteOperation` to do the flushing as you suggested. Additionally I worked out a first take at `ByteBufferReference`. The reader and writer indexes will help with the need in the future to have multiple framer decoders (ssl, followed by tcp).

With that, I think this is ready for you to take another look.

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java]: I think here we can go with an ordinary `BytesReference` and use it's efficient iterator `BytesRefIterator iterator()` the returned `BytesRef` is just a pointer into the underlying `byte[]` that you can wrap in with `ByteBuffer#wrap` and do your accounting what is left for writing in this method. so there is no need to use a `NetworkBytesReference`

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java]: here I'd do the same as above an pass in an some kind of BytesReference factory that can produce new BytesReferences and return a `CompositeBytesReference` instead the `int` to signal how much has been read. We can figure out how to do SSL and stuff afterwards this is too hard to do in one step

File: test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java]: I'd just don't worry about all the default read length and use `PagedBytesReference` and big arrays here. We can also do this in a followup and we just don't do any reuse etc. and simply / blindly always create a fresh byte array. also look at my suggestion above in `NioSocketChannel`

File: test/framework/src/main/java/org/elasticsearch/transport/nio/NetworkBytesReference.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/NetworkBytesReference.java]: I left some suggestions to remove this for now. I really don't want to block all the progress on this, lets try to work with `BytesReference` only. I left some suggestions in `NioSocketChannel`

File: test/framework/src/main/java/org/elasticsearch/transport/nio/CompositeByteBufferReference.java
Comment (s1monw) [inline test/framework/src/main/java/org/elasticsearch/transport/nio/CompositeByteBufferReference.java]: I left some suggestions to remove this for now. I really don't want to block all the progress on this, lets try to work with `BytesReference` only. I left some suggestions in `NioSocketChannel`

**Review by s1monw — CHANGES_REQUESTED** [review]
@tbrooks8  I love the progress here. I left some suggestions how to remove `NetworkBytesReference` for now and detach all this from the awesome progress we made here. Lets try to use `BytesReference` only and cut some corners so we can merge this. It's way way easier to iterate once this is in as a basis.

**Comment by tbrooks8** [comment]
@s1monw I made a number of changes to attempt to reflect your concerns.

I removed the composite reference I had introduced. And removed the abstract network buffer class. However, I left the `NetworkBytesReference`. Essentially - we still kind of need some type of reference that tracks how far we have written or read. I could use the raw `ByteBuffer`, but this would have require a lot of changes to the code and probably would have ended up being pretty complicated with a lot of test changes.

I did modify the `NetworkBytesReference` to be pretty simple. It takes a `BytesArray` instance for creation. And every `BytesReference` method is simply delegated to the `BytesArray`. The only thing that the `NetworkBytesReference` provides are indexes and access to `ByteBuffer` instances for usage with raw channels.

**Review by s1monw — APPROVED** [review]
@tbrooks8 I think this is a good compromise for now. I would like us to make use of bigarrays and paged bytes reference in the future but I think for now it's in good shape for an initial commit of this impl. We will be able to make way better progress once this is in and I thin we are ready to push this as a testing impl. 

There are tons of things we need to test, improve and optimize but lets get this in.

**Comment by tbrooks8** [comment]
Merged

**Pull request merged**
